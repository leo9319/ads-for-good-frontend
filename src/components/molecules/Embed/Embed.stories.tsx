import React from 'react';
import { Embed } from './Embed';
import type { Meta, StoryObj } from '@storybook/react';

const meta = {
  title: 'Library/Molecules/Embed',
  component: Embed,
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  decorators: [
    Story => (
      <div
        style={{ width: '100vw', height: '100vh', backgroundColor: 'white' }}
      >
        <Story />
      </div>
    ),
  ],
  argTypes: {
    src: {
      description: 'src of the embed',
      control: { type: 'text' },
    },
    size: {
      description: 'Size of the embed',
      control: 'select',
      options: ['1', '2', '3'],
    },
    className: {
      description: 'Additional class name for the embed',
      control: { type: 'text' },
    },
  },
  args: {
    size: '1',
  },
} satisfies Meta<typeof Embed>;

export default meta;
type Story = StoryObj<typeof meta>;

export const embed: Story = {
  args: {
    src: 'https://widget.tagembed.com/2169275',
  },
};
