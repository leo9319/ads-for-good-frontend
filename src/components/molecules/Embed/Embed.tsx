import AtomCard from '@components/atoms/AtomCard';
import { Skeleton } from '@radix-ui/themes';
import React, { ReactElement, useEffect, useRef } from 'react';

/**
 * @interface
 * The Embed props type of {@link Embed}.
 */
export type EmbedProps = {
  /**
   * The source URL of the embed/iframe.
   * @property
   */
  src: string;
  /**
   * The size of the card.
   * @defaultValue `1`
   * @property
   */
  size: '1' | '2' | '3';
  /**
   * Additional class name for the embed.
   * @property
   */
  className?: string;
};

/**
 *
 * @group Molecules
 * @category Component
 *
 * Embed component is a wrapper for an iframe element.
 * It can be used to embed external content such as socials feeds, etc.
 *
 * @param {EmbedProps} props - The props for the Embed component.
 *
 * @returns {JSX.Element} - The Embed component or null if no source is provided.
 *
 */
export const Embed = ({
  src,
  size = '1',
  className,
}: EmbedProps): ReactElement | null => {
  const ref = useRef<HTMLIFrameElement>(null);
  const [isLoading, setIsLoading] = React.useState(true);

  const handleLoad = () => {
    setIsLoading(false);
  };

  useEffect(() => {
    const iframe = ref.current;
    if (iframe) {
      iframe.addEventListener('load', handleLoad);
    }

    return () => {
      if (iframe) {
        iframe.removeEventListener('load', handleLoad);
      }
    };
  });

  if (!src) {
    return null;
  }

  return (
    <AtomCard size={size} className={className}>
      <Skeleton width="100%" height="100%" loading={isLoading}>
        <iframe
          ref={ref}
          src={src}
          style={{
            width: '100%',
            height: '100%',
            border: 'none',
          }}
          title="tagembed"
          loading="lazy"
        ></iframe>
      </Skeleton>
    </AtomCard>
  );
};

export default Embed;
