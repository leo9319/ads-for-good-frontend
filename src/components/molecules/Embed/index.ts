export { default } from './Embed';
export { Embed } from './Embed';
export type { EmbedProps } from './Embed';
