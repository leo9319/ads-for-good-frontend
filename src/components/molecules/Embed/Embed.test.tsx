import React from 'react';
import Embed from './Embed';
import { render } from '@testing-library/react';

describe('Embed', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <Embed src="https://widget.tagembed.com/2169076" size="1" />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
