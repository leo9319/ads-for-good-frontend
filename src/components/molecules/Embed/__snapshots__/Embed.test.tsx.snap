// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Embed should render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-jc-center rt-r-w rt-r-h wrapper"
      style="--width: 100%; --height: 100%;"
    >
      <div
        class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-jc-center rt-r-w card"
        style="--width: 100%; --cardPadding: 80px; --cardBorderRadius: 32px;"
      >
        <div
          class="rt-Flex rt-r-display-flex rt-r-ai-center rt-r-jc-center rt-r-w innerBox"
          style="--width: 100%;"
        >
          <iframe
            aria-hidden="true"
            class="rt-Skeleton rt-r-w rt-r-h"
            inert=""
            loading="lazy"
            src="https://widget.tagembed.com/2169076"
            style="--width: 100%; --height: 100%; width: 100%; height: 100%;"
            tabindex="-1"
            title="tagembed"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
